### PR TITLE
Ports gas reaction overhaul for SM

### LINKED ISSUE #768: Supermatter doesn't generate enouf heat fast enouf.
**Problem Description**:
The suermatter doesn't generate enouf heat fast enouf.

**What did you expect to happen**: 
I expectet the supermatter would power the ship by generating enouf heat.

**What happened instead**:
It       was   so      slow    heating      up.
**Why is this bad/What are the consequences:**
Because the supermatter is suposed to power the ship, good luck doing that with a -140 degrees hot loop.

**Steps to reproduce the problem**:
Set up the supermatter nothing fancy needed, get the cold loop running good, inject n2o into the hot loop, turn on emmiter, watch the power output and the heat in the hot loop climb slower then a snail climbing the Burj Khalifa.

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: It was the begining of the round.

### PR DESCRIPTION
**Pull request by ninjanomnom**
Ports https://github.com/tgstation/tgstation/pull/24785

Freon is not included and all references to it are removed.

This also includes partial ports of other features which are needed to complete the main port.

Tested and works without runtimes. Generated some power properly at least.

:cl: ninjanomnom
add: Most gases now have unique effects when surrounding the supermatter crystal.
tweak: The supermatter crystal can now take damage from too much energy and too much gas.
add: Added a dangerous overcharged state to the supermatter crystal.
add: Readded explosion delaminations, a new tesla delamination and allowed the singulo delamination to absorb the supermatter.
tweak: The type of delamination now depends on the state of the supermatter crystal.
tweak: Various supermatter engine rebalancing and fixes.
/:cl:

Fixes #768

### CODE REVIEW

**Comment by AdamElTablawy** [comment]
For clarification, the tesla delamination makes the supermatter decay into the tesla when it goes off rather than an explosion.

Essentially, the engine will kill everyone and everything rather than being barely survivable.

Not sure if you have the 'tesla explodes electronics' here too.

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -4,6 +4,21 @@
 
 #define in_range(source, user) (get_dist(source, user) <= 1)
 
+//Turfs
```
Comment (ike709) [inline code/__DEFINES/is_helpers.dm]: >porting the isturf helpers without converting all of the existing istype checks to isturf checks

Comment (ninjanomnom) [inline code/__DEFINES/is_helpers.dm]: yeah, I only did partial ports for some things just to get the gas interactions working

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -1567,4 +1567,4 @@ proc/pick_closest_path(value)
 	world.Export("http://[global.bot_ip]/?[list2params(data)]")
 
 /proc/to_chat(target, message)
-  target << message
```
Comment (ike709) [inline code/__HELPERS/unsorted.dm]: Were you using spaces instead of a tab...?

Comment (ninjanomnom) [inline code/__HELPERS/unsorted.dm]: No it's the browser editor fucking things up which I forgot not to use

**Comment by ike709** [comment]
inb4 the new engineering meta is to build singularity and tesla containment around the SM in case of delamination

**Comment by ike709** [comment]
Also high pressure windows aren't mapped in?

**Comment by ike709** [comment]
Line 415 of window.dm on tg has
var/ratio = obj_integrity / max_integrity

we have
var/ratio = health / max_health

So I'm kinda confused why this PR messes with object integrity when we don't have the obj damage refactor PR

File: code/game/objects/objs.dm [hunk 3]
```diff
@@ -9,6 +9,12 @@
 
 	var/damtype = "brute"
 	var/force = 0
+	
+	var/obj_integrity = 500
```
Comment (ike709) [inline code/game/objects/objs.dm]: the partial port pains me

Comment (ninjanomnom) [inline code/game/objects/objs.dm]: partial ports are fine, not everything has to be done in one go

Comment (Vivalas) [inline code/game/objects/objs.dm]: Something something atomic commits / PRs.

**Comment by ninjanomnom** [comment]
We'll get this in the rebase

**Pull request closed without merging**